### PR TITLE
Exercise 1: Todo-App

### PR DESCRIPTION
**Pull request by krisashy**

### CODE REVIEW

**Comment by Penomatikus** [comment]
@roschaefer Erbitten code review. Die Kommentare wollten wir _gerne_ drinnen lassen, auch wenn wir wissen, dass diese eigentlich zu entfernen sind.

File: VueJS/components/ListItem.js [hunk 1]
```diff
@@ -1 +1,61 @@
...
+        return {
+            editMode: false, 
+            newMessage: ""
+        }
+    },
+    template: `
+        <form id="test">
+            <li v-if="editMode == true" >                
+                <input type="text" placeholder="enter new value" v-on:input="updateMessage"> 
+                <button v-on:click="saveMe(todo)" type="button">
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:13]:
```suggestion
                <button @click="saveMe(todo)" type="button">
```
Use shorthands.

File: VueJS/components/ListItem.js [hunk 2]
```diff
@@ -1 +1,61 @@
...
+            this.$emit('delete-todo', todo.id)
+        },
+
+        cancelMe: function(event) {
+            this.editMode = false;
+        }, 
+
+        updateMessage: function(event) {
+            this.newMessage = event.target.value
+        }
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:59]: I think we have `v-model` for this: https://vuejs.org/v2/guide/forms.html#v-model-with-Components

File: VueJS/index.js [hunk 3]
```diff
@@ -3,11 +3,37 @@ import ListItem from './components/ListItem.js'
...
+    todos:  [
       { id: '1', message: 'Foo', },
       { id: '2', message: 'Bar', },
       { id: '3', message: 'Baz', }
     ]
-  }
+  },
+  methods: {
+    newTodo: function() {      
+      this.lastId++
```
Comment (roschaefer) [inline VueJS/index.js:20]: This is the 4th time I see this and it does not get better over time. :stuck_out_tongue: try to reduce state that needs to be updated. Remember:

> State is the root of all evil

What happens if you add items to the initial list of `todos`? Yes, you need to update this counter. This is a potential bug in the future.

File: VueJS/index.js [hunk 4]
```diff
@@ -3,11 +3,37 @@ import ListItem from './components/ListItem.js'
...
+      this.todos.push({id : this.lastId, message: ""})
+      // console.log("New Todo: [" + this.lastId + "]")     
+    },  
+    updateTodo: function(passedTodo) {
+        // console.log("(Großeltern) \nIch erhielt die todo [" + passedTodo.id + " | " + passedTodo.message + "] via $emit ")
+        // console.log("(Großeltern) \nFinde den entsprechenden index passend zur ID(" + passedTodo.id +") in todos[]")
+        let index = this.todos.findIndex( todo => todo.id == passedTodo.id)
```
Comment (roschaefer) [inline VueJS/index.js:27]:
```suggestion
        let index = this.todos.findIndex( todo => todo.id === passedTodo.id)
```
Always use triple equation for comparisons like this: https://codeburst.io/javascript-double-equals-vs-triple-equals-61d4ce5a121a

File: VueJS/index.js [hunk 5]
```diff
@@ -3,11 +3,37 @@ import ListItem from './components/ListItem.js'
...
+        // console.log("(Großeltern) \nUpdate das entsprechend gefundene Objekt in todos[]")
+        this.todos[index] = passedTodo
```
Comment (roschaefer) [inline VueJS/index.js:29]: First time I see sb. replacing the whole object, not only the property of the object. Great! :clap: 

You can do a one-liner here:
```js
this.todos = this.todos.map(t => t.id === passedTodo.id ? passedTodo : t)
```

File: VueJS/index.html [hunk 6]
```diff
@@ -1,8 +1,27 @@
...
+            <p>{{ message }} </p>
+            <p>{{ something }} </p>
+        </div>
 
-<div id="app">
-<!-- render your list here -->
-</div>
+        <div id="app">
+            <list
+                v-bind:todos="todos"
```
Comment (roschaefer) [inline VueJS/index.html:14]:
```suggestion
               :todos="todos"
```
Use shorthands

File: VueJS/components/ListItem.js [hunk 7]
```diff
@@ -1 +1,61 @@
...
+    props: ['todo'], 
+    data: () => {
+        return {
+            editMode: false, 
+            newMessage: ""
+        }
+    },
+    template: `
+        <form id="test">
+            <li v-if="editMode == true" >
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:11]:
```suggestion
            <li v-if="editMode">                
```

File: VueJS/components/ListItem.js [hunk 8]
```diff
@@ -1 +1,61 @@
...
+                </button>
+                <button v-on:click="deleteMe(todo)" type="button">
+                    Delete
+                </button>
+
+            </li>
+        </form> 
+    `,
+    methods: {
+        editMe: function(event) {
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:36]:
```suggestion
        editMe() {
```

File: VueJS/components/ListItem.js [hunk 9]
```diff
@@ -1 +1,61 @@
...
+            this.todo.message = this.newMessage
+            this.$emit('update-todo', todo)
+            this.editMode = false
+        },
+
+        deleteMe: function(todo) {
+            this.$emit('delete-todo', todo.id)
+        },
+
+        cancelMe: function(event) {
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:53]:
```suggestion
        cancelMe() {
```

File: VueJS/components/ListItem.js [hunk 10]
```diff
@@ -1 +1,61 @@
...
+
+            </li>
+        </form> 
+    `,
+    methods: {
+        editMe: function(event) {
+            this.editMode = true;
+        },
+        
+        saveMe: function(todo) {
```
Comment (roschaefer) [inline VueJS/components/ListItem.js:40]:
```suggestion
        saveMe(todo) {
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+# Quick installation guide
+1. Install VueJS DevTools:  
+ [Chrome extention](https://github.com/vuejs/vue-devtools#vue-devtools) or    
+ [Firefox extention](https://addons.mozilla.org/en-US/firefox/addon/vue-js-devtools/) or  
+ [Stand alone](https://github.com/vuejs/vue-devtools/blob/master/shells/electron/README.md)  
+ 
+2. Install npm und Node.js:  
+ [Win / Mac / Unix: Node.js/npm](https://nodejs.org/en/download/current/)  
+ 
+3. Install Visual Studio Code (if not installed with Node.js/nmp package automatically)
```
Comment (roschaefer) [inline README.md:20]: No, Visual Studio Code does not need to be installed. :stuck_out_tongue:

File: README.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+Folgende Studenten nehmen als Gruppe an dieser Übung teil:   
+544764 
+535285  
+549906  
+  
+# Installation guide  
+https://vuejs.org/v2/guide/installation.html 
+
+# Quick installation guide
+1. Install VueJS DevTools:
```
Comment (roschaefer) [inline README.md:12]:
```suggestion
1. Install VueJS DevTools (optionally):  
```

**Review by roschaefer — APPROVED** [review]
Yes, looks great! You passed exercise #1 congratulations!

congrats

This is my suggestion for the UX:

Screenshot - 2019-10-28T233800 316

**Comment by roschaefer** [comment]
@Penomatikus can you create a PR in your fork over here next time: https://github.com/Penomatikus/Systems-Development-and-Frameworks/pulls

Then you can also merge this into your own `master`. I'm going to close this, because your code shouldn't become part of the original repo.

Also:

Screenshot - 2019-10-28T232639 836

**Pull request closed without merging**